### PR TITLE
Expand/Collapse for folders in DiffTreeViewer

### PR DESCRIPTION
**Pull request by DenisUngemach**
The default behavior for double click on folders is to expand resp.
collapse them.

Fix for: https://github.com/eclipse-platform/eclipse.platform.ui/issues/1838

### CODE REVIEW

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java [hunk 1]
```diff
@@ -670,6 +673,35 @@ private void internalSetSelection(TreeItem ti, boolean fireOpen) {
...
 	}
 
+	private void expandCollapseAction(DoubleClickEvent event) {
+
+		ISelection s = event.getSelection();
+
+		if (!s.isEmpty()) {
+			Object e = getElement(s);
+
+			if (e instanceof DiffNode) {
```
Comment (BeckerWdf) [inline team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java:683]: Since Java 16 you can shorten this to:

```suggestion
			if (e instanceof DiffNode d) {
```
and then remove the 
```
DiffNode d = (DiffNode) e;
```
line

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java [hunk 2]
```diff
@@ -670,6 +673,35 @@ private void internalSetSelection(TreeItem ti, boolean fireOpen) {
...
+						collapseToLevel(e, 1);
+					else
+						expandToLevel(e, 1);
+				}
+			}
+		}
+	}
+
+	private static Object getElement(ISelection selection) {
+		if (selection instanceof IStructuredSelection) {
```
Comment (BeckerWdf) [inline team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java:697]: Same here:

```suggestion
		if (selection instanceof IStructuredSelection ss) {
```

**Review by BeckerWdf — CHANGES_REQUESTED** [review]

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java [hunk 3]
```diff
@@ -670,6 +673,35 @@ private void internalSetSelection(TreeItem ti, boolean fireOpen) {
 		}
 	}
 
+	private void expandCollapseAction(DoubleClickEvent event) {
+
+		ISelection s = event.getSelection();
+
+		if (!s.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (DenisUngemach) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java]: Done

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java (hunk 2 above)
Comment (DenisUngemach) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java]: Done

File: team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java (hunk 1 above)
Comment (DenisUngemach) [reply team/bundles/org.eclipse.compare/compare/org/eclipse/compare/structuremergeviewer/DiffTreeViewer.java]: Done

**Comment by DenisUngemach** [comment]
Thanks :-)

**Review by DenisUngemach — COMMENTED** [review]
Code has been changed.

**Review by BeckerWdf — APPROVED** [review]

**Comment by BeckerWdf** [comment]
I tested this on macOS. Works fine.
